File: app/src/main/java/club/hutcwp/lifeutil/ui/base/BaseActivity.java
package club.hutcwp.lifeutil.ui.base;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import androidx.annotation.LayoutRes;
import com.google.android.material.snackbar.Snackbar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import android.view.MenuItem;
import android.view.View;
import android.widget.Toast;

import club.hutcwp.lifeutil.R;
import club.hutcwp.lifeutil.app.AppGlobal;

/**
 * Created by hutcwp on 2017/4/15.
 * Mail : hutcwp@foxmail.com
 * Blog : hutcwp.club
 * GitHub : github.com/hutcwp
 */

public abstract class BaseActivity extends AppCompatActivity {

    protected Toolbar toolbar;

    @LayoutRes
    protected abstract int getLayoutId();

    protected abstract void initViews(Bundle savedInstanceState);

    protected abstract void loadData();

    /**
     * 初始化ToorBar
     */
    private void initToolBar() {
        toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        initTheme();
        setContentView(getLayoutId());
        initToolBar();
        initViews(savedInstanceState);
        loadData();
    }

    // 给左上角图标的左边加上一个返回的按钮
    protected void setDisplayHomeAsUpEnabled(boolean enable) {
        if (getSupportActionBar() != null)
            getSupportActionBar().setDisplayHomeAsUpEnabled(enable);
    }

    /**
     * 菜单创建
     * @param item 菜单Item
     * @return
     */
    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == android.R.id.home) {
            onBackPressed();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    /**
     * 设置主题
     */
    private void initTheme() {
        SharedPreferences pf = getSharedPreferences(AppGlobal.INSTANCE.getFILE_NAME(), Context.MODE_PRIVATE);
        int themeIndex =  pf.getInt("theme",0);
        switch (themeIndex) {
            case 0:
                setTheme(R.style.LapisBlueTheme);
                break;
            case 1:
                setTheme(R.style.PaleDogwoodTheme);
                break;
            case 2:
                setTheme(R.style.GreeneryTheme);
                break;
            case 3:
                setTheme(R.style.PrimroseYellowTheme);
                break;
            case 4:
                setTheme(R.style.FlameTheme);
                break;
            case 5:
                setTheme(R.style.IslandParadiseTheme);
                break;
            case 6:
                setTheme(R.style.KaleTheme);
                break;
            case 7:
                setTheme(R.style.PinkYarrowTheme);
                break;
            case 8:
                setTheme(R.style.NiagaraTheme);
                break;

        }
    }

    /**
     * showSnack
     *
     * @param s 要显示的内容
     */
    public void showSnack(View view , String s) {
        Snackbar snackbar = Snackbar.make(view, s, Snackbar.LENGTH_LONG);
        snackbar.show();
    }


}

File: read/src/main/java/club/hutcwp/lifeutil/ui/base/BaseActivity.kt
package club.hutcwp.lifeutil.ui.base

import android.content.Context
import android.content.SharedPreferences
import android.os.Bundle
import android.view.MenuItem
import android.view.View

import com.google.android.material.snackbar.Snackbar

import androidx.annotation.LayoutRes
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import club.hutcwp.lifeutil.R
import club.hutcwp.lifeutil.app.AppGlobal

/**
 * Created by hutcwp on 2017/4/15.
 * Mail : hutcwp@foxmail.com
 * Blog : hutcwp.club
 * GitHub : github.com/hutcwp
 */

abstract class BaseActivity : AppCompatActivity() {


    @get:LayoutRes
    protected abstract val layoutId: Int

    protected abstract fun initViews(savedInstanceState: Bundle?)

    protected abstract fun loadData()


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        initTheme()
        setContentView(layoutId)
        initViews(savedInstanceState)
        loadData()
    }

    // 给左上角图标的左边加上一个返回的按钮
    protected fun setDisplayHomeAsUpEnabled(enable: Boolean) {
        if (supportActionBar != null)
            supportActionBar!!.setDisplayHomeAsUpEnabled(enable)
    }

    /**
     * 菜单创建
     * @param item 菜单Item
     * @return
     */
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == android.R.id.home) {
            onBackPressed()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    /**
     * 设置主题
     */
    private fun initTheme() {
        val pf = getSharedPreferences(AppGlobal.FILE_NAME, Context.MODE_PRIVATE)
        val themeIndex = pf.getInt("theme", 0)
        when (themeIndex) {
            0 -> setTheme(R.style.LapisBlueTheme)
            1 -> setTheme(R.style.PaleDogwoodTheme)
            2 -> setTheme(R.style.GreeneryTheme)
            3 -> setTheme(R.style.PrimroseYellowTheme)
            4 -> setTheme(R.style.FlameTheme)
            5 -> setTheme(R.style.IslandParadiseTheme)
            6 -> setTheme(R.style.KaleTheme)
            7 -> setTheme(R.style.PinkYarrowTheme)
            8 -> setTheme(R.style.NiagaraTheme)
        }
    }

    /**
     * showSnack
     *
     * @param s 要显示的内容
     */
    fun showSnack(view: View, s: String) {
        val snackbar = Snackbar.make(view, s, Snackbar.LENGTH_LONG)
        snackbar.show()
    }
}
